lib/spack/spack/modules/lmod.py

```diff
@@ -65,6 +65,42 @@ def make_context(spec):
     return LmodContext(conf)
 
 
+def guess_core_compilers(store=False):
+    """Guesses the list of core compilers installed in the system.
+
+    Args:
+        store (bool): if True writes the core compilers to the
+            modules.yaml configuration file
+
+    Returns:
+        List of core compilers, if found, or None
+    """
+    core_compilers = []
+    for compiler_config in spack.compilers.all_compilers_config():
+        try:
+            compiler = compiler_config['compiler']
+            # A compiler is considered to be a core compiler if any of the
+            # C, C++ or Fortran compilers reside in a system directory
+            is_system_compiler = any(
+                os.path.dirname(x) in spack.util.environment.system_dirs
+                for x in compiler['paths'].values() if x is not None
+            )
+            if is_system_compiler:
+                core_compilers.append(str(compiler['spec']))
+        except (KeyError, TypeError, AttributeError):
+            continue
+
+    if store and core_compilers:
+        # If we asked to store core compilers, update the entry
+        # at 'site' scope (i.e. within the directory hierarchy
+        # of Spack itself)
+        modules_cfg = spack.config.get('modules', scope='site')
+        modules_cfg.setdefault('lmod', {})['core_compilers'] = core_compilers
+        spack.config.set('modules', modules_cfg, scope='site')
```

Comment:
This makes me nervous, especially since LLNL is a site where the core compilers are going to depend on the platform (we share a filesystem between clusters).

I can see why you'd want to write the site config for modules by default, but it's not consistent with other places in Spack where we write configs.  I guess this is a fundamental policy question -- where should we modify things by default?

The other places (compilers, repos, mirrors) where we write out configs write to the `spack.cmd.default_modify_scope()` by default, but they also allow the user to say where they want things written.  Could we integrate this with `spack module refresh` or some other command to give the user a bit more control?  I worry that people might be surprised if this location is written by default.

Reply:
I think both ideas are good, but let's postpone the second one.  I think we'll want/need that one for other reasons, too, but I haven't fully thought it out yet.

If the code that modifies the config file lives in `modules` then maybe the default config scope method needs to live in `config.py` instead of buried in `cmd/__init__.py`.

Reply:
@tgamblin Fine with you if I move `spack.cmd.default_modify_scope()` into `spack.config`? I'd like to avoid circular dependencies for `spack.modules` if I can (modules should be used by `spack.cmd.*` and not require it).